operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/InformerEventSource.java

```diff
@@ -154,19 +158,28 @@ public void onDelete(R resource, boolean b) {
   private synchronized void onAddOrUpdate(Operation operation, R newObject, R oldObject,
       Runnable superOnOp) {
     var resourceID = ResourceID.fromResource(newObject);
-    if (eventRecorder.isRecordingFor(resourceID)) {
-      log.debug("Recording event for: {}", resourceID);
-      eventRecorder.recordEvent(newObject);
-      return;
+
+    String previous = newObject.getMetadata().getAnnotations()
+        .get(KubernetesDependentResource.PREVIOUS_ANNOTATION_KEY);
+    boolean known = false;
+    if (previous != null) {
+      String[] parts = previous.split(",");
+      if (id.equals(parts[0])) {
+        if (oldObject == null && parts.length == 1) {
+          known = true;
+        } else if (oldObject != null && parts.length == 2
+            && oldObject.getMetadata().getResourceVersion().equals(parts[1])) {
```

Comment:
Hmm, maybe I'm not following, but this shouldn't be `newObject` here what is compared with previous. So that we know that we just received has the same version as in the annotation (own). Basically saying that if the resource version did not change since we updated the resource we can skip.

Reply:
> I don't think there's a great way to overcome that with local knowledge.

What I mean by that is that there is missing information from the event.  Just like onDelete provides unknownFinalState, onUpdate/onAdd could convey whether they are from a relist such that we know the prior state isn't exact.

Reply:
Yes. theoretically we could cache v2 version , and if the oldVersion == annotation version but newObjectVersion != v2  ==> emit the event

Reply:
> Yes. theoretically we could cache v2 version (we know that from update response) , and if the oldVersion == annotation version but newObjectVersion != v2 ==> emit the event

I was thinking along the same lines as well.  That will greatly narrow, but not eliminate the corner case.  I've updated the code to reflect this.

Reply:
The remaining - exceedingly narrow case where this doesn't work:

- starting with the temporary cache item as null, and informer cache state at v1
- a full relist of the informer needs to be initiated - just a restart of the watch shouldn't cause a problem
- the dependent resource initiates an update of v1, which will result in v2 - but the blocking call does not complete yet
- on the api server in response to v2, v3 is generated by some other actor
- the relist is processed such that it starts at v3, and emits an add event where the old state is v1 and the new is v3.  The logic will detect this as "known", since it matches state expectations - temporay cache item is null and v1 matches the annotation.

Reply:
> The remaining - exceedingly narrow case where this doesn't work:
> 
> * starting with the temporary cache item as null, and informer cache state at v1
> * a full relist of the informer needs to be initiated - just a restart of the watch shouldn't cause a problem
> * the dependent resource initiates an update of v1, which will result in v2 - but the blocking call does not complete yet
> * on the api server in response to v2, v3 is generated by some other actor
> * the relist is processed such that it starts at v3, and emits an add event where the old state is v1 and the new is v3.  The logic will detect this as "known", since it matches state expectations - temporary cache item is null and v1 matches the annotation.

This is what I mean to cover by this:

> Yes. theoretically we could cache v2 version (we know that from update response) , and if the oldVersion == annotation version but newObjectVersion != v2 ==> emit the event

Reply:
> Yes. theoretically we could cache v2 version (we know that from update response)

You won't cache that because the update call is completing at the end of the above sequence and by that time the informer cache has already moved ahead to v3.

Reply:
I mean caching it explicitly from the response.

Reply:
there would be still a minor chance I guess that we receive v3 before the v2 is cached, but that is extremely low. (Probably won't happen in our lifetime :) )

Reply:
> there would be still a minor chance I guess that we receive v3 before the v2 is cached, but that is extremely low. (Probably won't happen in our lifetime :) )

Yes this is what I mean by extremely narrow.